### ISSUE TITLE
Cross-compiled Nix references build platform binaries

### ISSUE THREAD

**Issue by cript0nauta** [NixOS/nixpkgs#75324]
**Describe the bug**
Building the cross-compiled version of Nix (I'm using an x86_64 system and want to use Nix in armv7l) with the following command succeeds:

`nix-build . -A pkgsCross.armv7l-hf-multiplatform.nixUnstable`

But it references the x86_64 versions of bash, coreutils and a few more derivations. This makes using Nix in my ARMv7l system fail because it will try to run x86_64 binaries.

**To Reproduce**
1. Run `nix-build . -A pkgsCross.armv7l-hf-multiplatform.nixUnstable` inside the nixpkgs directory (tested with `master` and `release-19.09` branches`)
2. Run `nix-store -q --references result | grep -E 'bash|coreutils'`. The printed outputs won't have `armv7l-unknown-linux-gnueabihf` in their name, indicating they are x86_64 derivations

**Expected behavior**
Rebuilding and running `nix-store -q --references result | grep -E 'bash|coreutils'` should print outputs with `armv7l-unknown-linux-gnueabihf` in their name.

**Workaround**

I was able to fix the bug with an ugly hack:

```diff
diff --git a/pkgs/tools/package-management/nix/default.nix b/pkgs/tools/package-management/nix/default.nix
index 5c55ac31588..afcb5303da7 100644
--- a/pkgs/tools/package-management/nix/default.nix
+++ b/pkgs/tools/package-management/nix/default.nix
@@ -16,6 +16,7 @@ common =
   , storeDir
   , stateDir
   , confDir
+  , coreutils, bash, gzip, gnutar
   , withLibseccomp ? lib.any (lib.meta.platformMatch stdenv.hostPlatform) libseccomp.meta.platforms, libseccomp
   , withAWS ? stdenv.isLinux || stdenv.isDarwin, aws-sdk-cpp
 
@@ -59,6 +60,18 @@ common =
       # Seems to be required when using std::atomic with 64-bit types
       NIX_LDFLAGS = lib.optionalString (stdenv.hostPlatform.system == "armv6l-linux") "-latomic";
 
+      preBuild = ''
+        coreutils="${coreutils}/bin" \
+          bash="${bash}/bin/bash" \
+          bzip2="${bzip2}/bin/bzip2" \
+          gzip="${gzip}/bin/gzip" \
+          xz="${xz}/bin/xz" \
+          tar="${gnutar}/bin/tar" \
+          tr="${coreutils}/bin/tr" \
+          substituteAllInPlace corepkgs/config.nix.in
+      '';
+
       preConfigure =
         # Copy libboost_context so we don't get all of Boost in our closure.
         # https://github.com/NixOS/nixpkgs/issues/45462
@@ -80,6 +93,7 @@ common =
           "--sysconfdir=${confDir}"
           "--disable-init-state"
           "--enable-gc"
+          "--with-coreutils-bin=${coreutils}/bin"
         ]
         ++ lib.optionals (!is20) [
           "--with-dbi=${perlPackages.DBI}/${perl.libPrefix}"
```

It built a working derivation of Nix, and I have been able to install and run Nix in my ARMv7l system (it is an Odroid HC2).

The `--with-coreutils-bin` option fixes the `result/etc/profile.d/nix.sh` file referencing an x86_64 version of coreutils. I haven't tested this change with other architectures, but this seems to be an acceptable change.

The `substituteAllInPlace` part is a horrible hack (that's why I didn't open a PR) that fixes the references to x86_64 derivations in the `result/share/nix/corepkgs/config.nix` file. There should be a better way to fix this, but I'm not sure how :(

**Metadata**

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.80, NixOS, 19.09.1510.131ff6dd9c8 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(root): `"nixos-19.09.1510.131ff6dd9c8"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
    - nix
    - nixUnstable
# a list of nixos modules affected by the problem
module:
```

**Comment by matthewbauer** [comment]
This was removed in an unreleased version of Nix here: https://github.com/NixOS/nix/commit/6a9c815734de3549a5116796830bd4ebfd006025. Unfortunately adding rust comes with its own issues in cross compiling.

**Comment by cript0nauta** [comment]
This appears to be fixed by @eamsden in #81317. It looks like #81153 was a duplicate of this issue. Unfortunately I can't confirm the issue is fixed because building `pkgsCross.armv7l-hf-multiplatform.nixUnstable` in master fails because of rustc.

If I cherry-pick the changes of #81317 into commit 131ff6dd9c8 (I used it when I originally reported the issue) and resolve the conflicts, `pkgsCross.armv7l-hf-multiplatform.nixUnstable` builds successfully but still contains a reference to x86_64 coreutils.

Although I think this issue should be fixed, it isn't important to me anymore. I already have a running armv7l machine with Nix, so now I build everything using it instead of cross-compiling.
